src/build/plan.rs

```diff
@@ -273,23 +287,34 @@ impl Plan {
         }
     }
 
-    pub fn prepare_work<T: AsRef<Path> + fmt::Debug>(&self, modified: &[T]) -> WorkStatus {
-        if !self.is_ready() {
-            return WorkStatus::NeedsCargo;
+    pub fn prepare_work<T: AsRef<Path> + fmt::Debug>(&mut self, manifest_path: &Path, modified: &[T], requested_cargo: bool) -> WorkStatus {
+        if self.package_map.is_none() || requested_cargo {
+            self.package_map = Some(PackageMap::new(manifest_path));
```

Comment:
iirc `cargo_metadata` executes `cargo` and parses JSON output from stdout - if we're already bundling Cargo in-process, I wonder if the resulting mapping couldn't be fetched already using `Workspace` discovery or during Cargo's `compile_ws` call?

Reply:
Yeah, I'm pretty sure we could compute this ourselves or get it out of Cargo somehow, but there didn't seem to be an obvious way, and `cargo metadata` is easy :-) I'll file an issue to follow up.